### PR TITLE
Updated text to be more clear about exceptions

### LINKED ISSUE #28: Client configuration and HTTP exceptions
The spec and doc-blocks refers to "client configuration" without defining what is meant by that.

I summise it has to do with the following clause, which I regard as an inconsistency in the specification.

> all HTTP clients should follow the Liskov substitutions principle. This means that all clients should act the same when sending a request.

So far, so good - but then:

> *By default* a HTTP client should not follow redirect nor throw exceptions on HTTP responses with status 4xx or 5xx.

"By *default*" - so in other words, you *are* making an allowance here for breaking LSP, perhaps with the understanding that this be done via configuration?

And again in the doc-block of `HttpException`:

> This exception MUST NOT be thrown when using the client's default configuration.

It sounds like you're making an allowance for alternate configuration that may cause this exception to be thrown after all?

If that's what is implied, I strongly oppose such an allowance - clients will *not* be interchangible, and will violate LSP, if we can't know whether a client might throw an `HttpException` or not.

This will cause severe problems for any library class that depends on `HttpClientInterface`, since, as an implementor of such a class, I can't know whether I might need to handle this class of exceptions or not.

For example, any HTTP API service-wrapper *will* need to know whether it should handle exceptions or not, and, effectively, this means you can only assume you *have* to handle it, and, in fact, you will need to always make sure you handle *both* - for example:

```php
class IceCreamClient
{
    private $http;
    public function __construct(HttpClientInterface $http) {
        $this->http = $http;
    }
    public function getIceCream($flavor) {
        try {
            $response = $this->http->sendRequest(new Request("http://icecream.test/flavor/{$flavor}"));
        } catch (HttpException $e) {
            if ($e->getResponse()->getStatusCode() === 404) {
                // handle flavor not found....
            }
        }
        if ($response->getStatusCode() === 404) {
            // also handle flavor not found...
        }
        // handle success...
    }
}
```

In other words, all client code must handle two different possible means of HTTP error-code reporting.

This is particularly harmful to clients of REST HTTP APIs where the API overloads HTTP status-codes with domain meaning, such as is (unfortunately, IMO, but never the less) very widespread and popular practice.

I'm suspect you may be trying to "please everyone" by making these allowances (?) but in the end, this will more likely end up harming everyone.

I think you need to make a firm decision and stand by it - either:

1. Stipulate of client implementations that they *must* throw the `HttpException`, or
2. Stipulate that *no* client implementation may throw an `HttpException`

As argued, if the answer is (1) then this will lead to awkward implementations and inconsistent logic and flow in clients such as illustrated by the example above, as (sadly) many popular APIs overload the meaning of HTTP status codes such as 404 in particular.

I would therefore strongly prefer (2) in which `HttpException` is merely included as utility (interface or perhaps even class) that clients (such as the one in the example above) can elect to throw if *they* encounter HTTP status codes that *they* weren't expecting.

In other words, we (as an implementor of `HttpClientInterface`) can't know or assume what the meaning of any HTTP status code is - it is too often completely dependent on the HTTP API in question.

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #28 

FYI @mindplay-dk, @sagikazarmark, @dbu, @kelunik

No exceptions should be thrown bases on HTTP codes. 

### Note
The php-http's PluginClient will violate the interface if configured with the ErrorPlugin. 

What do we think Guzzle will do? If the user uses an option to throw exception, should Guzzle throw `\InvalidArgument` or should it violate the interface?

The points here are:
* The interface says NEVER throw exception for HTTP status codes. 
* All consuming libraries should assume this behaviour. 
* We cannot stop HTTP clients to violate the interface.

### CODE REVIEW

File: proposed/http-client/http-client-meta.md [hunk 1]
```diff
@@ -41,8 +40,7 @@ specify the default behaviours.
 The intention of this PSR is ensure library developers that all HTTP clients have the same 
 **default behavior**. That means that all HTTP clients MUST follow Liskov substitution principle
 when no configuration is provided. The PSR does not try to restrict nor define configuration for 
-HTTP clients. An implementing library is free to be configured by the application author to follow
-redirects, to throw exceptions or any other possible setting.  
+HTTP clients.
```
Comment (kelunik) [inline proposed/http-client/http-client-meta.md:43]: This might be an issue. Consumers can't know whether redirects are followed or not and might fail if the client is configured to not follow any redirects.

File: proposed/http-client/http-client.md [hunk 2]
```diff
@@ -21,9 +21,9 @@ implementations. This would make libraries more stable since the reduced number
 dependencies and the likelihood to get in version conflicts is reduced.
 
 The second goal is that all HTTP clients should follow the [Liskov substitutions principle][Liskov].
-This means that all clients should act the same when sending a request. By default a HTTP client
-should not follow redirect nor throw exceptions on HTTP responses with status 4xx or 5xx.
-
+This means that all clients should act the same when sending a request. A consuming
+library SHOULD expect a HTTP client to NOT follow redirects NOR throw exceptions based
+on HTTP response statuses.
```
Comment (kelunik) [inline proposed/http-client/http-client.md:26]: With that language it's fine, but IMO a client should follow redirects by default.

File: proposed/http-client/http-client.md [hunk 3]
```diff
@@ -43,10 +43,9 @@ interface HttpClient
     /**
      * Sends a PSR-7 request.
      *
-     * If a request is sent without any prior configuration, an exception MUST NOT be thrown
-     * when a response is recieved, no matter the HTTP status code.
+     * An exception MUST NOT be thrown when a response is recieved, no matter the HTTP status code.
```
Comment (kelunik) [inline proposed/http-client/http-client.md:46]: This should probably contain: `when a valid response is received`.

File: proposed/http-client/http-client.md [hunk 4]
```diff
@@ -43,10 +43,9 @@ interface HttpClient
...
      *
-     * If a request is sent without any prior configuration, a HTTP client MUST NOT follow redirects.
+     * The HTTP client MUST NOT follow redirects.
```
Comment (kelunik) [inline proposed/http-client/http-client.md:48]: That makes the client a lot harder to use without breaking on redirects. Is there any reason to disable redirects per default?

File: proposed/http-client/http-client.md [hunk 5]
```diff
@@ -148,7 +147,8 @@ use Psr\Http\Client\Exception;
  * Thrown when a response was received but the request itself failed.
  *
  * This exception MAY be thrown on HTTP response codes 4xx and 5xx.
- * This exception MUST NOT be thrown when using the client's default configuration.
+ * A comsuming library SHOULD NOT except this exception to be thrown. The
+ * exception exists for middlewares, third party library and application code.
```
Comment (kelunik) [inline proposed/http-client/http-client.md:151]: This doesn't make any sense to me. Either it can be thrown, then consumers should expect it, or it can't be, then it can be deleted.

File: proposed/http-client/http-client-meta.md (hunk 1 above)
Comment (Nyholm) [reply proposed/http-client/http-client-meta.md:43]: Consumers will have to check the response status code and take the appropriate action.

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (Nyholm) [reply proposed/http-client/http-client.md:26]: No, why should 301, 302, 307(?) responses get special treatment?

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (Nyholm) [reply proposed/http-client/http-client.md:46]: What do you mean with "valid"? =)

**Comment by Nyholm** [comment]
Can we vote on this comment? Should the HttpException be deleted? 👍 / 👎

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:26]: Because they're defined that way in the RFC?

Comment (Nyholm) [reply proposed/http-client/http-client.md:26]: My argument is that the HTTP client should not try to be smart and special treat ANY response. It should simply send a a request and deliver the response. It should not care if it is 100, 200, 404 or 301.

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:46]: e.g. a malformed response

Comment (Nyholm) [reply proposed/http-client/http-client.md:46]: Can you give an example of a "malformed" response?
The way I see it; either you get a response form the server or you do not. If you do not, then a Network exception should be thrown.

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:26]: Good that you mention 1XX responses, because they're the ones that definitely need special treatment, see https://tools.ietf.org/html/rfc7231#section-6.2.

File: proposed/http-client/http-client-meta.md (hunk 1 above)
Comment (kelunik) [reply proposed/http-client/http-client-meta.md:43]: Which makes this component quite low-level, but fine.

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (Nyholm) [reply proposed/http-client/http-client.md:26]: Let me get this clear, if we special treat 1xx, 301 and 302 responses like you suggest, that would mean that a user of the HttpClientInterface never will receive a PSR response with those status codes. Is that what you are arguing for?

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:46]: Let's say you get an `HTTP/0.9` response, then you get a response, but do not consider that valid. With the current language you must not throw an exception in that case.

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:26]: Right, a user of the interface should never see a 1XX response. 3XX responses can be made configurable with an options array or flag in the method.

Comment (Nyholm) [reply proposed/http-client/http-client.md:26]: What if Im writing a small script to test my server behaviour? The script should make sure the server respond with 100 or 301. Shouldn't my script be able to use this PSR?

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (Nyholm) [reply proposed/http-client/http-client.md:46]: Yes, correct. We should not throw an exception in **that** case. That is still a valid version number. It is not officially recognised but it is valid. 

However, if the server responds with something that cannot be parsed as HTTP we should throw an exception.

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:26]: No. A 100 response is never a final HTTP response.

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:46]: IMO an `HTTP/0.9` response should definitely throw an exception. In fact a HTTP/0.9 response is just a response body as defined in https://www.ietf.org/rfc/rfc1945.txt. If a HTTP/1.x response claims to be any version other than 1.x, it should definitely result in an exception.

Comment (dbu) [reply proposed/http-client/http-client.md:46]: another example where i would expect an exception is with malformed headers. and we could explicitly mention connection problems. those should throw an exception, right?

File: proposed/http-client/http-client.md (hunk 4 above)
Comment (dbu) [reply proposed/http-client/http-client.md:48]: this is inconsistent with what we say above. and i would strongly opt to either MUST or MUST NOT redirects. 

while i prefer to not follow redirects in my applications, so that i can detect mistakes in my API clients (after all, if my client is redirected, it probably is redirected each time i do a request, which is very poor resource consumption) we should probably opt for the more forgiving choice of following redirects.

i feel that its fine to do allow configuration options to disable redirects in the client for when you want strict control (or in my case probably a plugin for the PluginClient to throw an exception when it encounters a 3xx state, so that i have to debug my problem)

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (dbu) [reply proposed/http-client/http-client.md:26]: using SHOULD here is not ok. these things have to be MUST or MUST NOT (i elaborated below)

File: proposed/http-client/http-client.md (hunk 4 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:48]: I guess you probably don't want an exception, but instead log the redirect, so it can be reviewed?

With an exception you instantly break the app that is supposed to continue working with the redirect.

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (mindplay-dk) [reply proposed/http-client/http-client.md:26]: IMO, the difference in behavior (follow vs not follow) is problematic for exactly the same reason that throwing vs not throwing for certain HTTP status codes was problematic: it means that the client can *not* be safely substituted for another client implementation.

In other words, consuming code will *have* to assume that clients do *not* follow redirects, because they might not - in exactly the same way that clients would *have* to assume that all implementations throw.

I think this behavior needs to be consistent.

I'm not sure which I would prefer though, following or not following redirects. There are use-cases for both. Likely the most common use-case is following - but stipulating that clients *must* follow redirects means that the client abstraction can't be used at all for those use-cases where it *shouldn't* follow redirects.

So it becomes a question of supporting all use-cases (not following) vs convenience in the majority of use-cases (following).

That's a tough call. Likely this is why clients at large make this configurable.

Comment (kelunik) [reply proposed/http-client/http-client.md:26]: Right, so the only sane behavior is not following unless we want to introduce some per-request configuration.

Following redirects brings quite some challenges that could be very opinionated. It's probably better to just use a library to wrap the calls if following is desired.

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (kelunik) [reply proposed/http-client/http-client.md:46]: Actually it continues down to truncated responses (connection closed before complete body received) and invalid transport encoding like invalid gzip or invalid chunked encoding sizes.

File: proposed/http-client/http-client-meta.md [hunk 6]
```diff
@@ -29,8 +29,7 @@ client will allow libraries to be decoupled from an implementation such as Guzzl
 ### Non-Goals
 
 * The purpose of this PSR is not to support asynchronous HTTP clients.
-* This PSR will not include how to configure a HTTP client. It does only
-specify the default behaviours.
+* This PSR will not include how to configure a HTTP client.
```
Comment (sagikazarmark) [inline proposed/http-client/http-client-meta.md]: does not

File: proposed/http-client/http-client.md (hunk 2 above)
Comment (sagikazarmark) [reply proposed/http-client/http-client.md:26]: Technically it shouldn't matter to a consuming code if the client followed redirections because it will receive a response anyway.

On the other hand if the client does not follow redirections, the consuming code needs to deal with it, which can actually be a valid use case too. For example: based on the redirected host new Authorization header needs to be added, etc.

I think it's okay not to follow redirections by default, because most of the cases the person who writes the consuming code also knows if it should expect a redirection or not.

Comment (mindplay-dk) [reply proposed/http-client/http-client.md:26]: Because this trade-off is ugly either way, I'm leaning towards suggesting a second argument be added, e.g. signature as follows:

    sendRequest(RequestInterface $request, bool $follow = true): ResponseInterface

This way, the default behavior is appropriate for the majority of use-cases - and for rare cases where the client needs to know if there was an intermediary redirect, they can set `$follow` to `false`.

I'm afraid there's no other way to satisfy the most common use-case without precluding the other less-common use-cases.

If we have to manually implement following redirects in almost every piece of client-code, anything that depends on this abstraction is going to be error-prone and will need tests and mocks for redirect handling, which would be pretty awkward.

If you're opposed to having this as an option, I'd say clients *should* follow redirects, and consumers will need to use a different client for those less-common cases where they need to know if there's a redirect. The majority of clients probably don't need to know, e.g. those fetching pages, data or assets/images - likely only special clients like spiders and diagnostic tools would need to know if there's a redirect.

What doesn't make sense, is to design for the minority use-case.

Comment (mindplay-dk) [reply proposed/http-client/http-client.md:26]: Alternatively:

```php
const NO_FOLLOW = 1;
const CHECK_STATUS = 2;

public function sendRequest(RequestInterface $request, int $options = 0): ResponseInterface;
```

This way, consumers can elect for each use-case if they want to disable following, whether they want to map HTTP status-codes to exceptions, and possible other options added in the future.

I know that adding a bunch of options is a slippery road to start down, but both of these options seems to support common enough use-cases - consumer code will be a little more verbose in some cases, but it will be a lot more readable and easier to understand, since you don't need to know about configuration state being carried around by the client; and it respects LSP.

Comment (dbu) [reply proposed/http-client/http-client.md:26]: i am strictly -1 on a `bool $follow` argument. if anything, then flags, or an array of key-value that can be handled with eg. the symfony OptionsResolver.

but i am not happy to open this can of worms. there is a ton of things that could or could not be configured at runtime. it also implies that each client MUST be capable of doing those things (error checking, redirecting, ...whatever else we then want to add). reusable libraries would then be littered with configuration to each call where it should be configuration of the client behaviour. 

imo libraries should either use the low level client directly, or wrap it in something like the php-http PluginClient to make it handle redirects / throw exceptions (although those should be converted to domain exceptions actually, not http exceptions)...

Comment (mindplay-dk) [reply proposed/http-client/http-client.md:26]: > i am not happy to open this can of worms

Me neither, but...

> reusable libraries would then be littered with configuration to each call where it should be configuration of the client behaviour

I strongly disagree that things like following redirects or (especially) throwing errors "should be" configuration of the client. In my opinion, at least these two options in particular, describe *not* how the client behaves, but how the individual request is processed.

As we've seen and discussed above, client behavior, when described as the state of the client, breaks SRP and effectively ends up forcing every consumer to create their own instance.

I think that's completely unnecessary. Any piece of consumer code, such as a REST client, that needs a particular consistent behavior from this method, can simply wrap the call in an internal method and supply the required options - which they would likely do anyway, to avoid manually constructing requests everywhere, adding authentication headers, etc. for example:

```php
class SomeRestClient 
{
    public function __construct(HttpClient $client) {
        $this->client = $client;
    }
    private function request(RequestInterface $request) {
        return $this->client->sendRequest($request, HttpClient::CHECK_STATUS);
    }
}
```

In other cases, where a consumer is performing only a single request, it doesn't really matter how these options get configured - it's the same workload, and this makes the expected behavior clearly visible at the call site, rather than needing to read through bootstrapping code elsewhere.

If you insist on request behavior options as state in the client, you'd need a client-factory interface - but it leads to the same end result, just more clutter and less transparency, so I'd be :-1: on that.

Yes, having a set of defined options *does* imply that each client implementation must be capable of doing certain things - isn't that the whole point of standardizing this in the first place though?

I'm not suggesting we brainstorm for an endless variety of "nice to have" options we can add, but I think, from the discussion above, it's pretty clear that there are common, real use-cases for the two options we've discussed here.

The client wouldn't be very useful without these, as indicated by your suggestion that decorators/wrappers could provide this behavior - but then, in my opinion, we're really talking about a client abstraction that's too low-level to be of any real-world use. If the abstraction provides only a low-level client, and every use-case needs some proprietary "client wrapper" to implement the major behavior, what have we won, really? How is a non-shared dependency on "client wrappers" any better than a non-shared dependency on the client itself?

For a client abstraction to succeed, it has to be useful in real use-cases - otherwise the whole thing seems like a waste of time.

**Comment by Nyholm** [comment]
There has been some good discussions here. Most not related though, I've created separate issues for the topic that came up.

File: proposed/http-client/http-client.md (hunk 3 above)
Comment (dbu) [reply proposed/http-client/http-client.md:46]: invalid checksum probably as well. can we rephrase this to be clear? lets insert "valid Response" - imo together with "no matter the HTTP status code" it should be clear that a 4xx or 5xx is not allowed to trigger an exception..

File: proposed/http-client/http-client.md (hunk 5 above)
Comment (dbu) [reply proposed/http-client/http-client.md:151]: => #41

**Comment by dbu** [comment]
i think apart from the discussion about exceptions for non-valid responses this is ready to merge. it is clarifications of the existing spec. lets continue the discussion in the other issues.
